keps/sig-storage/3476-volume-group-snapshot/README.md

```diff
@@ -0,0 +1,1660 @@
+
+## Proposal for VolumeGroupSnapshot
+
+This proposal introduces new CRDs VolumeGroupSnapshot, VolumeGroupSnapshotContent, and VolumeGroupSnapshotClass.
+
+### Create VolumeGroupSnapshot
+
+A VolumeGroupSnapshot can be created from multiple PVCs with a label on the PVCs specified by the labelSelector in the VolumeGroupSnapshot if the CSI driver supports the CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT capability.
+
+#### Dynamic provisioning
+
+* Admin creates a VolumeGroupSnapshotClass.
+* User creates a VolumeGroupSnapshot with label selector that matches the label applied to all PVCs to be snapshotted together.
+* This will trigger the VolumeGroupSnapshot controller to create a VolumeGroupSnapshotContent API object, and also call the CreateVolumeGroupSnapshot CSI function. It will also create multiple VolumeSnapshot API objects with volumeGroupSnapshotName in the status and the corresponding VolumeSnapshotContents with the snapshot handle. The VolumeSnapshot and VolumeSnapshotContent will point to each other before these objects are created in the API server to avoid triggering the VolumeSnapshot controller to create new individual objects. The CSI snapshotter sidecar will not call CSI driver in this case. If needed, GetVolumeGroupSnapshot CSI function will be called to retrieve individual snapshot statuses until all snapshots are ready to use.
+* CreateVolumeGroupSnapshot CSI function response
+  * The CreateVolumeGroupSnapshot CSI function should return a list of snapshots (Snapshot message defined in CSI Spec) in its response. The VolumeGroupSnapshot controller can use the returned list of snapshots to construct corresponding individual VolumeSnapshotContents and VolumeSnapshots, wait for VolumeSnapshots and VolumeSnapshotContents to be bound, and update SnapshotList in the VolumeGroupSnapshot Status and SnapshotContentList in the VolumeGroupSnapshotContent Status.
+
+apiVersion: snapshot.storage.k8s.io/v1
+```
+kind: VolumeSnapshot
+metadata:
+  name: snapshot1
+spec:
+  source:
+    persistentVolumeClaimName: vsc1
+status:
+  volumeGroupSnapshotName: groupSnapshot1
+```
+
+* An admissions controller or finalizer should be added to prevent an individual snapshot from being deleted that belongs to a VolumeGroupSnapshot.
+* In the CSI spec, it is specified that it is required for individual snapshots to be returned along with the group snapshot.
+
+#### Pre-provisioned VolumeGroupSnapshot
+
+Admin can create a VolumeGroupSnapshotContent, specifying an existing VolumeGroupSnapshotHandle in the storage system and specifying a VolumeGroupSnapshot name and namespace. Then the user creates a VolumeGroupSnapshot that points to the VolumeGroupSnapshotContent name.
+
+Admin will retrieve all volumeSnapshotHandles in the Volume Group Snapshot from the storage system, create VolumeSnapshotContents pointing to the volumeSnapshotHandles. Then the user can create VolumeSnapshots pointing to the VolumeSnapshotContents.
+
+### Delete VolumeGroupSnapshot
```

Comment:
Since individual VolumeSnapshots are created by the GroupSnapshot logic at creation time, the deletion should also be handled by the GroupSnapshot logic.  We will not be calling CSI DeleteSnapshot RPCs.

Reply:
I added more clarifications.

Reply:
https://github.com/kubernetes/enhancements/pull/1551/commits/e1bcf82bd5f175f121bc8745b9e6fe87bddc3092